File: Tree/Easy/Lnt242-Convert-Binary-Tree-to-Linked-Lists-by-Depth/solution_bfs.cpp
/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    /**
     * @param root the root of binary tree
     * @return a lists of linked list
     */
    vector<ListNode*> binaryTreeToLists(TreeNode* root) {
        // bfs + level order traversal
        // tc:O(n); sc:O(n)
        if(!root) {
            return {};
        }
        vector<ListNode*> res;
        queue<TreeNode*> queue;
        queue.push(root);
        while(queue.size() > 0) {
            int levelSize = queue.size();
            ListNode* dummy = new ListNode(-1);
            ListNode* cur = dummy;
            for(int i = 0; i < levelSize; i++) {
                TreeNode* node = queue.front();
                queue.pop();
                cur->next = new ListNode(node->val);
                cur = cur->next;
                if(node->left) {
                    queue.push(node->left);
                }
                if(node->right) {
                    queue.push(node->right);
                }
            }
            res.push_back(dummy->next);
        }
        return res;
    }
};

File: BFS/Medium/199-Binary-Tree-Right-Side-View/solution_bfs.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public IList<int> RightSideView(TreeNode root) {
        // bfs + level order traversal
        // tc:O(n); sc:O(n)
        if(root == null) {
            return new List<int>();
        }
        IList<int> res = new List<int>();
        Queue<TreeNode> queue = new Queue<TreeNode>();
        queue.Enqueue(root);
        while(queue.Count > 0) {
            int levelSize = queue.Count;
            for(int i = 0; i < levelSize; i++) {
                TreeNode node = queue.Dequeue();
                if(i == 0) {
                    res.Add(node.val);
                }
                if(node.right != null) {
                    queue.Enqueue(node.right);
                }
                if(node.left != null) {
                    queue.Enqueue(node.left);
                }
            }
        }
        return res;
    }
}